perfkitbenchmarker/providers/azurestack/azurestack_virtual_machine.py

```diff
@@ -0,0 +1,446 @@
+         '--output', 'json',
+         '--name', self.name] + self.resource_group.args)
+    try:
+      json.loads(stdout)
+      return True
+    except:
+      return False
+
+  def GetInternalIP(self):
+    """Grab some data."""
+
+    stdout, _ = vm_util.IssueRetryableCommand(
+        [azurestack.AZURESTACK_PATH, 'network', 'nic', 'show',
+         '--output', 'json',
+         '--name', self.name] + self.resource_group.args)
+
+    response = json.loads(stdout)
+    return response['ipConfigurations'][0]['privateIpAddress']
+
+  def _Delete(self):
+    self._deleted = True
+
+
+class AzureStackVirtualMachineMetaClass(resource.AutoRegisterResourceMeta):
+  """Metaclass for AzureStackVirtualMachine.
+
+  Registers default image pattern for each operating system.
+  """
+
+  def __init__(self, name, bases, dct):
+    super(AzureStackVirtualMachineMetaClass, self).__init__(name, bases, dct)
+    if hasattr(self, 'OS_TYPE'):
+      assert self.OS_TYPE, '{0} did not override OS_TYPE'.format(self.__name__)
+      assert self.IMAGE_URN, (
+          '{0} did not override IMAGE_URN'.format(self.__name__))
+
+
+class AzureStackVirtualMachine(virtual_machine.BaseVirtualMachine):
+  """Object representing an AzureStack Virtual Machine."""
+
+  __metaclass__ = AzureStackVirtualMachineMetaClass
+  CLOUD = providers.AZURESTACK
+  # Subclasses should override IMAGE_URN.
+  IMAGE_URN = None
+
+  def __init__(self, vm_spec):
+    """Initialize an AzureStack virtual machine.
+
+    Args:
+      vm_spec: virtual_machine.BaseVmSpec object of the vm.
+    """
+    super(AzureStackVirtualMachine, self).__init__(vm_spec)
+    self.network = azurestack_network.AzureStackNetwork.GetNetwork(self)
+    self.firewall = azurestack_network.AzureStackFirewall.GetFirewall()
+    self.max_local_disks = 1
+    self._lun_counter = itertools.count()
+    self._deleted = False
+
+    self.resource_group = azurestack_network.GetResourceGroup()
+    self.public_ip = AzureStackPublicIPAddress(self.zone, self.name + '-public-ip')
+    self.nic = AzureStackNIC(self.network.subnet,
+                        self.name + '-nic', self.public_ip.name,
+                        vm_spec.accelerated_networking)
+    self.storage_account = self.network.storage_account
+    self.image = vm_spec.image or self.IMAGE_URN
+
+    disk_spec = disk.BaseDiskSpec('azurestack_os_disk')
+    self.os_disk = azurestack_disk.AzureStackDisk(disk_spec,
+                                        self.name,
+                                        self.machine_type,
+                                        self.storage_account,
+                                        None,
+                                        is_image=True)
+
+  def _CreateDependencies(self):
+    """Create VM dependencies."""
+    self.public_ip.Create()
+    self.nic.Create()
+
+  def _Create(self):
+    create_cmd = (
+        [azurestack.AZURESTACK_PATH, 'vm', 'create', '--use-unmanaged-disk',
```

Comment:
There are very few changes here as compared to azure_virtual_machine. This is one such change where we have the extra argument --use-unmanged-disk.

We're going to have trouble keeping this in sync with azure_virtual_machine unless we're able to share more code, or make this inherit from azure virtual machine.

Reply:
That is correct that there isn’t too much in the way of change against the current Azure provider. Due to Azure Stacks nature of it being a scaled down version of public Azure, it does have its limitations. I say this in the ability to port command sets from Azure to Azure Stack. Currently Azure CLI expects a certain profile to be present when interacting with Azure Stack:

https://docs.microsoft.com/en-us/cli/azure/?view=azure-cli-2017-03-09-profile

Azure Stack also lacks the complete and full functionality of public Azure, so like you picked up on, the use of ‘managed disks’ is not present in Azure Stack. So this needs to be tweaked to explicitly state --use-unmanaged-disk, as you can appreciate there are many other commands that expect things in a different order. Or are simply not yet present. As Azure Stack catches up more and functions become available, I expect people to be able to contribute and update the AzureStack provider to include/remove as it matures.

Reply:
I agree with Steve that refactoring to share more code, probably inheriting from azure virtual machine and perhaps overriding unsupported functions is better than two parallel providers.

Reply:
Thanks @dlott, I agree this approach would be far better. Is this something you are able to fold in or compile to the current Azure provider?

Reply:
I'm not sure I follow your question, but folding this in to the current Azure provider seems reasonable. If there aren't too many non-trivial changes, could you add a flag to the Azure provider that would switch to using Azure Stack instead of Azure?

Reply:
Would you be able to provide an example of this, and I will look to amend accordingly?